### PR TITLE
feat(geometry): add projection validity and masked depth warping

### PR DESCRIPTION
**Pull request by NaCr05**
## What does this pull request do?

Adds `projection_valid_mask` for finite pinhole-camera XYZ with positive depth, and `warp_frame_depth_with_mask` for depth warping with explicit validity and zero-filled invalid samples. The new warp rejects originally invalid destination depths before transformation and uses source-image dimensions for normalization and strict pixel-center bounds. Existing `warp_frame_depth` behavior is preserved.

A detached candidate path determines validity before live arithmetic and sampling, preventing rejected inputs from contaminating gradients. The mask describes depth, projection and image-bound validity; it does not estimate occlusion visibility.

## Related issue or discussion

Related to #4555.

## What did you check?

Validation covers coordinates, invalid depths, translation, source/target size differences, bounds, gradients, low precision and existing warp compatibility. Focused CPU tests passed on PyTorch 2.5.1 and 2.14.0; Linux full-graph Inductor passed. The helper was scripted; fixed-shape trace, torch.export, legacy ONNX export and ONNX Runtime were checked, including changed validity after export. Full repository pre-commit, type checks and documentation build passed. Dynamic-shape export and CUDA/MPS are not claimed.

- Windows CPU, Python 3.14.6 / PyTorch 2.14.0: **179 passed** (focused tests plus the public API inventory).
- Linux CPU, Python 3.12.13 / PyTorch 2.5.1: **31 eager + 2 Inductor tests passed**.
- Both test configurations exercised float32 and float64. Counts are parameterized test instances.
- Existing depth-warp compatibility tests: **41 passed**; the float16/bfloat16-focused run: **31 passed**, including shared dtype-independent tests.

Commands used (run from each checkout with its own interpreter/environment):

```text
pixi run -e default pre-commit-all
pixi run -e default typecheck
pixi run -e default build-docs
pixi run -e default test-module tests/geometry/test_projection_validity.py tests/test_api_surface.py --device=cpu --dtype=float32,float64
# Linux, PyTorch 2.5.1 environment
python -m pytest tests/geometry/test_projection_validity.py --device=cpu --dtype=float32,float64
KORNIA_TEST_OPTIMIZER=inductor python -m pytest tests/geometry/test_projection_validity.py --device=cpu --dtype=float32,float64 -k dynamo
```

## How did AI help?

Codex assisted with the implementation, regression tests, documentation and local validation. Numerical behavior was checked against independent references or mathematical invariants, and the supported PyTorch floor was tested separately from the current CPU environment. The API scope was selected by the contributor; the validation above was run by Codex.

### CODE REVIEW

**Comment by ducha-aiki** [comment]
Thanks for the PR. The motivation here is real, but this lands ahead of the decision it depends on: #4555 is explicitly a discussion issue ("It asks for a decision; there is no fix to review yet") and it has no replies so far. Please take the design there first. If you want this to land, the arguments need to be made in #4555, and the PR can follow once the questions are settled.

The points this PR decides implicitly, and which #4555 leaves open:

1. **What "valid" means.** The PR fixes `z > eps` with an absolute `eps = 1e-8`. #4555 asks whether `eps` should be absolute (its meaning depends on scene units) or dtype-aware, and whether validity is a per-camera-model rule, since angular models such as `KannalaBrandtK3` legitimately project rays past 90°.
2. **Whether image bounds belong in the mask.** `warp_frame_depth_with_mask` folds depth validity, `z > eps` and in-image bounds into one mask. #4555 asks whether covisibility is part of validity or a separate mask.
3. **API shape for the warps.** The starting point in #4555 is a `return_mask` flag on the existing warps, defaulting to off. This PR instead adds a second public function, `warp_frame_depth_with_mask`, that re-implements the warp (~100 lines) with different semantics from `warp_frame_depth`: zero fill, a strict pixel-centre bounds rule, a float32 upcast for half inputs. That leaves two near-duplicate public warps to maintain and keep consistent. If you think a separate function is the better shape, please argue it in #4555.
4. **Warp defaults and scope.** Whether invalid pixels are zero-filled by default, and which call sites adopt the rule (`project_points`, `PinholeCamera.project`, `project_points_z1`, `cam2pixel`, the `kornia.sensors` models, `DepthWarper`, `warp_kpts` with its own `z + 1e-4`). This PR covers `warp_frame_depth` only, so the two warps in #4555's repro would still disagree on missing depth.
5. **Gradients.** The detached-candidate / safe-argument approach here looks like the right direction for question 5 in #4555. It is worth describing there, including how it behaves on torch 2.5.1.

`projection_valid_mask` on its own is close to option (a) in #4555 and would likely survive the discussion nearly unchanged, but its `eps` semantics and the per-camera-model question should be agreed first, because the name and signature become public API.

I am not reviewing the implementation further until #4555 has a conclusion. Converting this to a draft in the meantime would be appreciated.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
